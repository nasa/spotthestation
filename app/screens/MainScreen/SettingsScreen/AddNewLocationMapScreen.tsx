import { useNavigation } from "@react-navigation/native"
import { observer } from "mobx-react-lite"
import React, { useCallback, useEffect, useRef, useState } from "react"
import { ViewStyle, TextStyle, View, Pressable, TextInput } from "react-native"
import { useSafeAreaInsets } from "react-native-safe-area-context"
import Modal from "react-native-modal"
import { Icon, Screen, Text, Button } from "../../../components"
import { colors, spacing, typography } from "../../../theme"
import { IconLinkButton } from "../../OnboardingScreen/components/IconLinkButton"
import { translate } from "../../../i18n/translate"
import Snackbar from "react-native-snackbar"
import { LatLng } from "react-native-maps"
import { api, LocationType } from "../../../services/api"
import { MapBox } from "../components/MapBox"
import { useStores } from "../../../models"
import { StyleFn, useStyles } from "../../../utils/useStyles"
import { LocationAutocomplete } from "../../../components/LocationAutocomplete"

export const AddNewLocationMapScreen = observer(function AddNewLocationMapScreen() {
  const {
    $headerStyleOverride,
    $container,
    $topButtonsContainer,
    $topContainer,
    $button,
    $locations,
    $active,
    $locationsListContainer,
    $dropdownSelected,
    $locationsRow,
    $locationsRowText,
    $dropdownText,
    $modal,
    $buttonText,
    $modalBodyContainer,
    $contentContainer,
    $close,
    $title,
    $modalButton,
    $map,
  } = useStyles(styles)

  const navigation = useNavigation()
  const { setNewSavedLocation, savedLocations } = useStores()
  const topInset = useSafeAreaInsets().top

  const addressRef = useRef<TextInput>()
  const [isFocus, setIsFocus] = useState(false)
  const [isSave, setIsSave] = useState(false)
  const [textValue, setTextValue] = useState("")
  const [location, setLocation] = useState<LocationType>(null)
  const [marker, setMarker] = useState<LatLng>(null)

  const handleClear = () => {
    addressRef.current?.clear()
    setTextValue("")
    setLocation(null)
  }

  useEffect(() => {
    ;(async () => {
      if (marker) {
        const res = await api.reverseGeocode(marker.latitude, marker.longitude)
        if (res.kind !== "ok") return

        setLocation({
          location: { lat: marker.latitude, lng: marker.longitude },
          title: res.name || res.address || "Location",
          subtitle: res.address,
        })

        setTextValue(res.address)
        setIsSave(true)
      }
    })().catch((e) => console.log(e))
  }, [marker])

  const handleNavigate = () =>
    navigation.navigate("LocationSettings" as never, { update: Date.now() } as never)

  const handleSave = useCallback(() => {
    if (savedLocations.find((item) => item.title === location.title)) {
      Snackbar.show({
        text: translate("snackBar.locationExist"),
        duration: Snackbar.LENGTH_LONG,
        action: {
          text: translate("snackBar.ok"),
          textColor: "green",
          onPress: () => {
            Snackbar.dismiss()
          },
        },
      })
      return
    }

    setNewSavedLocation(location).catch((error) => {
      Snackbar.show({
        text: error,
        duration: Snackbar.LENGTH_LONG,
        action: {
          text: translate("snackBar.ok"),
          textColor: "green",
          onPress: () => {
            Snackbar.dismiss()
          },
        },
      })
    })

    Snackbar.show({
      text: translate("snackBar.locationSaved"),
      duration: Snackbar.LENGTH_LONG,
      action: {
        text: translate("snackBar.ok"),
        textColor: "green",
        onPress: () => {
          Snackbar.dismiss()
        },
      },
    })
    handleNavigate()
  }, [location])

  const headerStyle = { ...$headerStyleOverride }
  headerStyle.top = Number(headerStyle.top) + topInset

  return (
    <Screen
      preset="fixed"
      contentContainerStyle={$container}
      style={{ backgroundColor: colors.palette.neutral900 }}
      statusBarStyle="light"
      dismissKeyboardOnPress={false}
    >
      <MapBox
        style={$map}
        withNightOverlay={false}
        onPress={({ geometry }) => {
          setMarker({ latitude: geometry.coordinates[1], longitude: geometry.coordinates[0] })
        }}
        markers={marker ? [marker] : []}
        zoomEnabled
      />
      <View style={[$topContainer, headerStyle]}>
        <View style={$topButtonsContainer}>
          <IconLinkButton
            icon="x"
            buttonStyle={$button}
            iconColor={colors.palette.neutral250}
            iconSize={20}
            onPress={() => handleNavigate()}
          />
        </View>
        <LocationAutocomplete
          ref={addressRef}
          placeholder={translate(
            "settings.locationSettingsData.addNewLocation.searchInputPlaceholder",
          )}
          onPress={(data) => {
            setLocation({
              title: data.name,
              subtitle: data.display_name,
              location: { lat: Number(data.lat), lng: Number(data.lon) },
              sightings: [],
            })
            setMarker({
              latitude: Number(data.lat),
              longitude: Number(data.lon),
            })
            setIsSave(true)
          }}
          styles={{
            textInputContainer: isFocus ? [$locations, $active] : $locations,
            textInput: {
              ...$dropdownText,
              ...$dropdownSelected,
            },
            listView: $locationsListContainer,
            row: $locationsRow,
            separator: { height: 0 },
            container: { flex: 0 },
          }}
          textInputProps={{
            allowFontScaling: false,
            placeholderTextColor: colors.palette.neutral450,
            value: location?.subtitle || textValue,
            onFocus: () => setIsFocus(true),
            onBlur: () => setIsFocus(false),
            onChangeText: (text) => setTextValue(text),
            clearButtonMode: "never",
          }}
          renderRow={({ description }) => {
            return (
              <Text
                text={description}
                style={$locationsRowText}
                ellipsizeMode="tail"
                numberOfLines={1}
              />
            )
          }}
          renderLeftAccessory={({ style }) => (
            <Icon icon="pin" size={28} color={colors.palette.neutral450} style={style} />
          )}
          renderRightAccessory={({ style }) =>
            isFocus &&
            textValue && (
              <Icon
                icon="xCircle"
                size={28}
                color={colors.palette.neutral450}
                style={style}
                onPress={handleClear}
              />
            )
          }
        />
      </View>

      <Modal
        isVisible={isSave}
        onBackdropPress={() => setIsSave(!isSave)}
        onSwipeComplete={() => setIsSave(!isSave)}
        animationIn="slideInUp"
        animationOut="slideOutDown"
        swipeDirection="down"
        useNativeDriver
        useNativeDriverForBackdrop
        hideModalContentWhileAnimating
        propagateSwipe
        backdropOpacity={0.65}
        style={$modal}
      >
        <View style={$modalBodyContainer}>
          <Pressable style={$close} onPress={() => setIsSave(!isSave)}>
            <Icon icon="x" size={20} color={colors.palette.neutral450} />
          </Pressable>
          <View style={$contentContainer}>
            <Text text={location?.subtitle} style={$title} />
            <Button
              tx="settings.locationSettingsData.addNewLocation.confirnModalButton"
              pressedStyle={$modalButton}
              textStyle={$buttonText}
              style={$modalButton}
              onPressIn={handleSave}
            />
          </View>
        </View>
      </Modal>
    </Screen>
  )
})

const styles: StyleFn = ({ scale, fontSizes, lineHeights }) => {
  const $headerStyleOverride: TextStyle = {
    top: scale(18),
  }

  const $container: ViewStyle = {
    flex: 1,
    backgroundColor: colors.backgroundDark,
    height: "100%",
  }

  const $topButtonsContainer: ViewStyle = {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
    marginBottom: scale(36),
  }

  const $topContainer: ViewStyle = {
    position: "absolute",
    left: 0,
    top: 0,
    width: "100%",
    paddingHorizontal: scale(36),
  }

  const $button: ViewStyle = {
    backgroundColor: colors.palette.neutral550,
    width: scale(42),
    height: scale(42),
  }

  const $locations: ViewStyle = {
    borderWidth: scale(1.5),
    borderColor: "transparent",
    borderRadius: scale(28),
    height: scale(56),
    backgroundColor: colors.palette.neutral350,
    overflow: "hidden",
    marginBottom: scale(18),
  }

  const $active: ViewStyle = {
    borderWidth: scale(1.5),
    borderColor: colors.palette.buttonBlue,
    backgroundColor: colors.palette.neutral350,
  }

  const $locationsListContainer: ViewStyle = {
    borderRadius: scale(12),
    backgroundColor: colors.palette.neutral550,
    overflow: "hidden",
    width: "85%",
    alignSelf: "center",
    marginTop: scale(3),
  }

  const $dropdownLeftAccessory: ViewStyle = {
    marginStart: spacing.large,
    height: scale(56),
    justifyContent: "center",
    alignItems: "center",
  }

  const $dropdownRightAccessory: ViewStyle = {
    marginEnd: spacing.large,
    height: scale(56),
    justifyContent: "center",
    alignItems: "center",
  }

  const $dropdownSelected: TextStyle = {
    color: colors.palette.neutral250,
  }

  const $locationsRow: TextStyle = {
    backgroundColor: "transparent",
    paddingHorizontal: scale(spacing.large),
  }

  const $locationsRowText: TextStyle = {
    fontSize: fontSizes[18],
    lineHeight: lineHeights[22],
    color: colors.palette.neutral250,
  }

  const $dropdownText: TextStyle = {
    flex: 1,
    height: scale(56),
    alignSelf: "stretch",
    fontFamily: typography.primary.normal,
    fontSize: fontSizes[18],
    lineHeight: lineHeights[22],
    paddingHorizontal: 0,
    marginHorizontal: scale(spacing.small),
    textAlignVertical: "center",
    borderRadius: 0,
    backgroundColor: "transparent",
  }

  const $modal: ViewStyle = {
    flex: 1,
    justifyContent: "flex-end",
    left: 0,
    margin: 0,
  }

  const $buttonText: TextStyle = {
    color: colors.palette.neutral100,
    fontSize: fontSizes[18],
    fontFamily: typography.primary.medium,
    lineHeight: lineHeights[22],
  }

  const $modalBodyContainer: ViewStyle = {
    backgroundColor: colors.palette.neutral350,
    borderTopLeftRadius: scale(18),
    borderTopRightRadius: scale(18),
  }

  const $contentContainer: ViewStyle = {
    width: "100%",
    paddingHorizontal: scale(36),
    paddingBottom: scale(18),
    alignItems: "center",
    marginTop: scale(56),
  }

  const $close: ViewStyle = {
    position: "absolute",
    top: 0,
    right: 0,
    padding: scale(18),
  }

  const $title: TextStyle = {
    textAlign: "center",
    color: colors.palette.neutral450,
    fontSize: fontSizes[18],
    fontFamily: typography.primary.normal,
    lineHeight: lineHeights[22],
    paddingBottom: scale(24),
  }

  const $modalButton: ViewStyle = {
    width: "40%",
    height: scale(56),
    backgroundColor: colors.palette.buttonBlue,
    borderRadius: scale(28),
    borderWidth: 0,
    marginTop: scale(24),
    marginBottom: scale(24),
  }

  const $map: ViewStyle = { flex: 1 }

  return {
    $headerStyleOverride,
    $container,
    $topButtonsContainer,
    $topContainer,
    $button,
    $locations,
    $active,
    $locationsListContainer,
    $dropdownLeftAccessory,
    $dropdownRightAccessory,
    $dropdownSelected,
    $locationsRow,
    $locationsRowText,
    $dropdownText,
    $modal,
    $buttonText,
    $modalBodyContainer,
    $contentContainer,
    $close,
    $title,
    $modalButton,
    $map,
  }
}
