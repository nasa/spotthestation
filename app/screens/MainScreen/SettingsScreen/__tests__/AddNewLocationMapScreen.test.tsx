import React from "react"
import { AddNewLocationMapScreen } from "../AddNewLocationMapScreen"
import { NavigationContainer } from "@react-navigation/native"
import { act, fireEvent, render, waitFor } from "@testing-library/react-native"

it("renders correctly", async () => {
  const component = render(
    <NavigationContainer>
      <AddNewLocationMapScreen />
    </NavigationContainer>,
  )

  await act(async () => {
    await new Promise((resolve) => setTimeout(resolve, 100))
  })

  expect(component.toJSON()).toMatchSnapshot()
})

test("the text input", async () => {
  const { getByPlaceholderText } = render(
    <NavigationContainer>
      <AddNewLocationMapScreen />
    </NavigationContainer>,
  )

  const textInput = getByPlaceholderText("homeScreen.selectLocation.inputPlaceholder undefined")
  fireEvent.changeText(textInput, "Some location")

  await waitFor(() => {
    expect(textInput.props.value).toBe("Some location")
  })
})

test("closes the modal window after clicking the Close button", async () => {
  const { getByPlaceholderText, queryByText } = render(
    <NavigationContainer>
      <AddNewLocationMapScreen />
    </NavigationContainer>,
  )

  const textInput = getByPlaceholderText("homeScreen.selectLocation.inputPlaceholder undefined")
  fireEvent.changeText(textInput, "Some location")

  await waitFor(() => {
    expect(queryByText("Confirm Location")).toBeNull()
  })
})
