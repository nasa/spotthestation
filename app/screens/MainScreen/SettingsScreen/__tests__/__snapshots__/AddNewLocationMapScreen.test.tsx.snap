// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div />
  <View
    style={
      [
        {
          "left": 0,
          "paddingHorizontal": 52,
          "position": "absolute",
          "top": 0,
          "width": "100%",
        },
        {
          "top": 26,
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 52,
        }
      }
    >
      <View
        accessibilityRole="imagebutton"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          animatedStyle={
            {
              "value": {},
            }
          }
          collapsable={false}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#353739",
              "borderRadius": 143,
              "display": "flex",
              "height": 60,
              "justifyContent": "center",
              "overflow": "hidden",
              "position": "relative",
              "width": 60,
            }
          }
        >
          <View
            style={
              [
                {
                  "backgroundColor": "transparent",
                },
                [
                  {
                    "alignItems": "center",
                    "backgroundColor": "#1E2022",
                    "borderRadius": 143,
                    "display": "flex",
                    "height": 60,
                    "justifyContent": "center",
                    "overflow": "hidden",
                    "position": "relative",
                    "width": 60,
                  },
                  undefined,
                ],
              ]
            }
          >
            <ViewManagerAdapter_ExpoBlurView
              proxiedProperties={
                {
                  "intensity": 0,
                  "tint": "default",
                }
              }
              style={
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            />
            <View>
              <RCTImageView
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                internal_analyticTag={null}
                resizeMode="contain"
                source={
                  [
                    {
                      "height": undefined,
                      "uri": undefined,
                      "width": undefined,
                    },
                  ]
                }
                style={
                  {
                    "height": 29,
                    "overflow": "hidden",
                    "resizeMode": "contain",
                    "tintColor": "#DEDDDE",
                    "width": 29,
                  }
                }
                tintColor="#DEDDDE"
              />
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      collapsable={false}
      style={
        {
          "zIndex": 1,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            [
              {
                "alignItems": "flex-start",
                "backgroundColor": "#1E2022",
                "borderRadius": 40,
                "flexDirection": "row",
                "height": 80,
                "overflow": "hidden",
              },
              false,
              undefined,
              {
                "paddingStart": 0,
              },
              {
                "paddingEnd": 0,
              },
              {
                "backgroundColor": "#1E2022",
                "borderColor": "transparent",
                "borderRadius": 40,
                "borderWidth": 2,
                "height": 80,
                "marginBottom": 26,
                "overflow": "hidden",
              },
            ]
          }
        >
          <View>
            <RCTImageView
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              internal_analyticTag={null}
              resizeMode="contain"
              source={
                [
                  {
                    "height": undefined,
                    "uri": undefined,
                    "width": undefined,
                  },
                ]
              }
              style={
                {
                  "alignItems": "center",
                  "height": 80,
                  "justifyContent": "center",
                  "marginStart": 35,
                  "overflow": "hidden",
                  "resizeMode": "contain",
                  "tintColor": "#ADADAE",
                  "width": 40,
                }
              }
              tintColor="#ADADAE"
            />
          </View>
          <TextInput
            accessibilityHint="type for search location"
            accessibilityLabel="search location"
            accessibilityRole="search"
            accessible={true}
            allowFontScaling={false}
            autoCapitalize="none"
            autoCorrect={false}
            clearButtonMode="never"
            editable={true}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="homeScreen.selectLocation.inputPlaceholder undefined"
            placeholderTextColor="#ADADAE"
            style={
              [
                {
                  "alignSelf": "stretch",
                  "color": "#DEDDDE",
                  "flex": 1,
                  "fontFamily": "Helvetica Neue",
                  "fontSize": 26,
                  "height": 80,
                  "marginHorizontal": 17,
                  "paddingHorizontal": 0,
                  "paddingVertical": 0,
                  "textAlignVertical": "center",
                },
                false,
                undefined,
                undefined,
                {
                  "alignSelf": "stretch",
                  "backgroundColor": "transparent",
                  "borderRadius": 0,
                  "color": "#DEDDDE",
                  "flex": 1,
                  "fontFamily": "Helvetica Neue",
                  "fontSize": 26,
                  "height": 80,
                  "lineHeight": 32,
                  "marginHorizontal": 17,
                  "paddingHorizontal": 0,
                  "textAlignVertical": "center",
                },
              ]
            }
            textAlignVertical="top"
            underlineColorAndroid="rgba(0, 0, 0, 0)"
            value=""
          />
        </View>
      </View>
    </View>
  </View>
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={true}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    onSwipeComplete={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      [
        "portrait",
        "landscape",
      ]
    }
    swipeDirection="down"
    swipeThreshold={100}
    transparent={true}
    useNativeDriverForBackdrop={true}
    visible={false}
  />
</div>
`;
